Better parser error handling

NathanW2 opened the pull request:
## Description
Crappy parser messages are crappy so lets make this a little bit better for the user.  With this we can now highlight errors using line and column numbers in the expression builder.

This is still a work in progress but opened for feedback and code review.

Currently supports two highlight methods.
- Underline
- Marker

Underline is used for unknown functions (although we never really have them yet due to how the code is written), wrong args, or invalid args)

![image](https://user-images.githubusercontent.com/381660/38480867-8e2dd8d4-3bb7-11e8-9145-37d7866c254d.png)

Marker point is for every other error at a single location which may or may not have a friendly message yet. We can add better handling for this in the future but it's here mainly to show the location of the error in the parser.

![image](https://user-images.githubusercontent.com/381660/38480923-d44bb9e4-3bb7-11e8-85d0-97ba08569482.png)

I will also improve the error display so that we show the real error without having to click a new window which is a bit hidden.

Funded by my QGIS dev days at [TechnologyOne](https://www.technologyonecorp.com/)

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [ ] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [ ] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

nyalldawson commented on src/core/expression/qgsexpression.h:
```diff
@@ -115,6 +115,23 @@ class CORE_EXPORT QgsExpression
+      };
+
+      ParserErrorType errorType = ParserErrorType::Unknown;
+      int firstLine;
```
Initialize to 0 (or -1?) here

nyalldawson commented on src/core/expression/qgsexpression.h:
```diff
@@ -115,6 +115,23 @@ class CORE_EXPORT QgsExpression
     Q_DECLARE_TR_FUNCTIONS( QgsExpression )
   public:
 
+    struct CORE_EXPORT ParserError
```
Don't forget doxygen here

nyalldawson commented:
Awesome! Those `unexpected end` messages are the worst. This is lovely stuff

NathanW2 commented on tests/src/core/testqgsexpression.cpp:
```diff
@@ -257,6 +257,36 @@ class TestQgsExpression: public QObject
       QCOMPARE( !exp.hasParserError(), valid );
     }
 
+    void parsing_error_line_column_data()
```
@nyalldawson how do these look for you?

nyalldawson replied:
Beautiful

NathanW2 commented:
How do people feel about this?  I'm an running the tests again to see if they pass but ok to merge if they do?

nyalldawson commented on src/core/expression/qgsexpression.h:
```diff
@@ -115,6 +115,49 @@ class CORE_EXPORT QgsExpression
+      enum ParserErrorType
+      {
+        Unknown = 0,
+        FunctionUnknown = 1,
```
Can you please add docstrings for these values too?

NathanW2 replied:
Yep.

nyalldawson reviewed (commented):
I'm in favour. I noticed you used 3.0 as the version added - are you planning to backport this? (I'd be in favour of a backport, as this fixes a severe usability issue with expressions)

NathanW2 commented:
@nyalldawson ah yeah that was a copy paste error.  I can backport but I will leave that to others to make the call on.

DelazJ commented on src/core/qgsexpressionparser.yy:
```diff
@@ -193,22 +195,28 @@ expression:
-            exp_error(parser_ctx, "Function is not known");
+            QgsExpression::ParserError::ParserErrorType errorType = QgsExpression::ParserError::FunctionUnknown;
+            parser_ctx->parserError.errorType = errorType;
+            exp_error(&yyloc, parser_ctx, "Function is not known");
```
I don't know if this is the right place in the commit (sorry!) but could these messages be translatable. the `unexpected end` message is bad but it's really worse when you do not speak English (and we have a couple of places where error messages are only shown in English).

NathanW2 replied:
@DelazJ Yeah they should be.  I will look into that in another PR as I have lots of stuff to change in here at the moment so just doing in bits.

NathanW2 commented:
Yes they should be. Let me look into it.

On Thu, Apr 12, 2018 at 9:17 PM, Harrissou Sant-anna <
notifications@github.com> wrote:

> *@DelazJ* commented on this pull request.
> ------------------------------
>
> In src/core/qgsexpressionparser.yy
> <https://github.com/qgis/QGIS/pull/6767#discussion_r181045617>:
>
> > @@ -193,22 +195,28 @@ expression:
>            {
>              // this should not actually happen because already in lexer we check whether an identifier is a known function
>              // (if the name is not known the token is parsed as a column)
> -            exp_error(parser_ctx, "Function is not known");
> +            QgsExpression::ParserError::ParserErrorType errorType = QgsExpression::ParserError::FunctionUnknown;
> +            parser_ctx->parserError.errorType = errorType;
> +            exp_error(&yyloc, parser_ctx, "Function is not known");
>
> I don't know if this is the right place in the commit (sorry!) but could
> these messages be translatable. the unexpected end message is bad but
> it's really worse when you do not speak English (and we have a couple of
> places where error messages are only shown in English).
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/6767#pullrequestreview-111577061>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAXS3KSxOG7_7RGIOsh2ri8d-Pe2R5Hdks5tnzfZgaJpZM4TL91n>
> .
>

NathanW2 committed:
[BUGFIX][needs-docs] Allow expression parser to report better error
location.

We return the line and column to allow builder to highlight
that location for the user.
